CMake/Autodownload.cmake

```diff
@@ -0,0 +1,84 @@
+          list(FILTER DIRECTORIES EXCLUDE REGEX ".*\.tar\.xz")
+        endif ()
+        list(LENGTH DIRECTORIES DIRECTORIES_LEN)
+        message("Print directories: " ${DIRECTORIES})
+        if (DIRECTORIES_LEN GREATER 0)
+          if (${DEPS_NAME} MATCHES "armadillo")
+            list(GET DIRECTORIES 0 ARMADILLO_DIR)
+            set(ARMADILLO_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${ARMADILLO_DIR}/include" CACHE INTERNAL "")
+
+          elseif (${DEPS_NAME} MATCHES "ensmallen")
+            list(GET DIRECTORIES 0 ENSMALLEN_DIR)
+            set(ENSMALLEN_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${ENSMALLEN_DIR}/include" CACHE INTERNAL "")
+
+          elseif (${DEPS_NAME} MATCHES "cereal")
+            list(GET DIRECTORIES 0 CEREAL_DIR)
+            set(CEREAL_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${CEREAL_DIR}/include" CACHE INTERNAL "")
+
+          elseif(${DEPS_NAME} MATCHES "boost")
+            list(GET DIRECTORIES 0 Boost_DIR)
+            set(Boost_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${Boost_DIR}/" CACHE INTERNAL "")
+          
+          elseif(${DEPS_NAME} MATCHES "OpenBLAS")
+            list(GET DIRECTORIES 0 OPENBLAS_DIR)
+            execute_process(COMMAND make TARGET=ARMV8 BINARY=64 HOSTCC=gcc CC=${CMAKE_C_COMPILER} FC=${CMAKE_FORTRAN_COMPILER} NO_SHARED=1
+                            WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/${OPENBLAS_DIR})
+            set(OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/${OPENBLAS_DIR}/libopenblas.a" CACHE INTERNAL "")
+          endif()
+
+        else ()
+          message(FATAL_ERROR 
+                  "Problem unpacking ${DEPS_NAME}! Expected only one directory ${DEPS_NAME};. Try removing the directory ${CMAKE_BINARY_DIR}/deps and reconfiguring.")
+        endif ()
+      endif ()
+      if (${DEPS_NAME} MATCHES "stb")
+        set(STB_DIR "stb")
+        file(MAKE_DIRECTORY "${CMAKE_BINARY_DIR}/deps/stb")
+        if(PACKAGE MATCHES "stb_image.h")
+          execute_process(COMMAND mv "deps/stb_image.h" "deps/stb")
+        elseif(PACKAGE MATCHES "stb_image_write.h")
+          execute_process(COMMAND mv "deps/stb_image_write.h" "deps/stb")
+        endif()
+        if(EXISTS "${CMAKE_BINARY_DIR}/deps/${STB_DIR}/stb_image.h" AND EXISTS "${CMAKE_BINARY_DIR}/deps/${STB_DIR}/stb_image_write.h")
+          check_hash (http://mlpack.org/files/stb/hash.md5
+              "${CMAKE_BINARY_DIR}/deps/${STB_DIR}"
+              HASH_CHECK_FAIL)
+          if (HASH_CHECK_FAIL EQUAL 0)
+            set(STB_IMAGE_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${STB_DIR}" CACHE INTERNAL "")
+          else ()
+            message(WARNING
+                "stb/stb_image.h is not installed. Image utilities will not be available!")
+          endif()
+        endif()
+      endif()
+    else ()
+      list(GET DOWNLOAD_STATUS_LIST 1 DOWNLOAD_ERROR)
+      message(FATAL_ERROR
+          "Could not download armadillo! Error code ${DOWNLOAD_STATUS}: ${DOWNLOAD_ERROR}!  Error log: ${DOWNLOAD_LOG}")
```

Comment:
I guess we might need to include the package name as a parameter, so that we can format this error message correctly.